### PR TITLE
Add support for hidden annotations while using the in-text annotation notation

### LINKED ISSUE #3621: Hidden annotation with link in values not possible (SMW_PARSER_LINV)
### Setup and configuration

- SMW version: 3.0.0
- MW version: 1.31.1
- PHP version: 7.2.10-0ubuntu0.18.04.1
- DB system (MySQL, Blazegraph, etc.) and version: 5.7.24-0ubuntu0.18.04.1

- `$smwgParserFeatures = SMW_PARSER_LINV;`
```
{
    "namespace": "SMW_NS_PROPERTY",
    "page": "Has text",
    "contents": "[[Has type::Text]]"
}
```
### Issue
Hidden annotation with link in values is not possible:
`[[Has text::[[Main Page]]| ]]`
results to "Main Page|" instead of a hidden annotation.

`{{#set:Has text=[[Main Page]]}}` works as expected adding a hidden annotation.

### Questions:
1. If this is not an intended use case, should I update the docu https://www.semantic-mediawiki.org/wiki/Help:Links_in_values, https://www.semantic-mediawiki.org/wiki/Help:$smwgLinksInValues with a hint to use #set for hidden annotation?
2. Would SemanticMediaWiki/tests/phpunit/Integration/JSONScript/TestCases/p-0449.json be the right place to add this test case?

### PR DESCRIPTION
**Pull request by gesinn-it**
This PR is made in reference to: #3621 

This PR addresses or contains:
- fix provided by @mwjames (thx!)
- updated test case 

This PR includes:
- [x] Tests (unit/integration)
- [x] CI build passed

Fixes #3621

### CODE REVIEW

**Comment by mwjames** [comment]
I'm guessing you are not finished since the other mentioned test cases aren't tested yet.

```
Was is the expected annotation? Was it the expected output?

[[Has text::[[Has number::12]] [[Main Page]]| ]]
[[Has text::[[Has number::123| ]] [[Main Page]]| ]]
[[Has text::[[Has number::1234|Foo]] [[Main Page]]| ]]
[[Has text::[[Has number::12345| ]] [[Main Page]] ]]
```

**Comment by gesinn-it** [comment]
@mwjames oh, you are right, I just had `[[Has text::[[Has number::12]] [[Main Page]]| ]]` in mind...

I'll prepare an update to the PR.

**Comment by gesinn-it** [comment]
The parser works recursive, so `[[Has text::[[Has number::12]] [[Main Page]]| ]]` will be replaced by `[[Has text::12 Main Page| ]]` (after first parsing) with the effect, that the suggested fix will not detect (replace) the pattern `]]|.` I was thinking about a different pattern (like `| ]])` but this will not work, as the outer `[[` and `]]` has been removed in the pre-processing.

I suggest to add test cases that clearly tests the additional cases mentioned by @mwjames plus docu (@kghbln) to be added to the wiki.

**Comment by gesinn-it** [comment]
**Case `[[Has text::[[Has number::12]] [[Main Page]]| ]]`**:

First, the parser removes the outer brackets, `[[Has text::[[Has number::12]] [[Main Page]]| ]]` and processes the inner annotation, which internally leads to `Has text::12 &#x005B;&#x005B;Main Page&#x005D;&#x005D;&#124;`. The fix then replaces `&#x005D;&#x005D;&#124;` with `&#x005D;&#x005D;|` which in the end gives a hidden annotation `Has text` with a value of `12 [[Main Page]]`.

One more case we should think of:
* `[[Has text::[[Has number::12]] [[Main Page]] lorem| ]]` which leads to `12 Main Page lorem|`

File: src/Parser/LinksEncoder.php [hunk 1]
```diff
@@ -228,8 +228,13 @@ private static function replace( $match, $parser, $isOffAnnotation = false ) {
 			$match
 		) );
 
+        // Decode the pipe as part of a hidden marker (`...]]|`) to ensure that
```
Comment (mwjames) [inline src/Parser/LinksEncoder.php]: Seems we missing a tab here!

File: tests/phpunit/Integration/JSONScript/TestCases/p-0449.json [hunk 2]
```diff
@@ -33,6 +33,30 @@
...
+		},
+		{
+			"page": "Example/P0449/9",
+			"contents": "AAA[[Has text::[[Has number::42| ]] [[Lorem ipsum]] ]]BBB"
+		},
+		{
+			"page": "Example/P0449/10",
+			"contents": "AAA[[Has text::[[Has number::42]] [[Lorem ipsum]] Foo| ]]BBB"
+		},
 		{
```
Comment (mwjames) [inline tests/phpunit/Integration/JSONScript/TestCases/p-0449.json]: I suspect you wanted to add:

```
One more case we should think of:

    [[Has text::[[Has number::12]] [[Main Page]] lorem| ]] which leads to 12 Main Page lorem|
```

File: tests/phpunit/Integration/JSONScript/TestCases/p-0449.json [hunk 3]
```diff
@@ -134,6 +158,161 @@
...
+			"subject": "Example/P0449/10",
+			"assert-store": {
+				"semantic-data": {
+					"strictPropertyValueMatch": false,
+					"propertyCount": 4,
+					"propertyKeys": [
+						"_SKEY",
+						"_MDAT",
+						"Has text",
+                        "Has number"
```
Comment (mwjames) [inline tests/phpunit/Integration/JSONScript/TestCases/p-0449.json]: Formatting!

File: tests/phpunit/Integration/JSONScript/TestCases/p-0449.json [hunk 4]
```diff
@@ -134,6 +158,161 @@
...
+					],
+					"propertyValues": [
+                        "42",
+                        "42 [[Lorem ipsum]] Foo"
+                    ]
```
Comment (mwjames) [inline tests/phpunit/Integration/JSONScript/TestCases/p-0449.json]: Formatting!

**Comment by gesinn-it** [comment]
sorry for the mess with the formatting. My editor pulled an `.editorconfig` from parent directory.

How about adding `.editorconfig` in SMW's root folder to prevent such issues in future?

```
root = true

[*]
end_of_line = lf
charset = utf-8
indent_size = 4
indent_style = tab
insert_final_newline = false
trim_trailing_whitespace = false
```

File: tests/phpunit/Integration/JSONScript/TestCases/p-0449.json (hunk 2 above)
Comment (gesinn-it) [reply tests/phpunit/Integration/JSONScript/TestCases/p-0449.json]: yes, that's coverd by P0449/10

**Comment by mwjames** [comment]
Thanks for the contribution.

**Pull request merged**